### PR TITLE
mrc-558: restore support for lapply

### PR DESCRIPTION
**Pull request by richfitz**
Somewhat simpler than what was there before, but should work.

### CODE REVIEW

**Comment by r-ash** [comment]
https://github.com/mrc-ide/rrq/commit/ab9eb5078b638f39fb61d93543d0bbe2909d35c9

File: R/rrq_controller.R [hunk 1]
```diff
@@ -69,6 +69,36 @@ R6_rrq_controller <- R6::R6Class(
       task_submit(self$con, self$keys, dat, key_complete)
     },
 
+    ## One option for envir_base would be to use search()[2] (~=
+    ## parent.env(.GlobalEnv)) as that's going to be relatively likely
+    ## to be ok because that's where the packages end.
+    lapply = function(X, FUN, ..., DOTS = NULL,
+                      envir = parent.frame(), envir_base = NULL,
```
Comment (r-ash) [inline R/rrq_controller.R]: What's the `envir_base` for here? How is it different from `envir`?

Comment (r-ash) [reply R/rrq_controller.R]: Some documentation of what the args are for would be super helpful.

File: R/rrq_controller.R [hunk 2]
```diff
@@ -69,6 +69,36 @@ R6_rrq_controller <- R6::R6Class(
       task_submit(self$con, self$keys, dat, key_complete)
     },
 
+    ## One option for envir_base would be to use search()[2] (~=
+    ## parent.env(.GlobalEnv)) as that's going to be relatively likely
+    ## to be ok because that's where the packages end.
+    lapply = function(X, FUN, ..., DOTS = NULL,
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/rrq_controller.R]: This is now hopefully explained in the docs a bit. The main issue is that to program against this `DOTS` is sometimes needed, and sometimes a non-magic version is needed. Most of the time the user can ignore this complication though. We will care about this if we start having to parametrise the calls we make to enqueue

File: R/bulk.R [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+## Try to work out what version of a function we are likely to have
+## remotely.  Some of this could nodoubt be done with rlang, but with
+## ~400 functions in that package and the crazy rate at which they
+## deprecate and change I thin that would be more work in the long
+## run.  This used to work quite well with lazyeval but that also
+## changed behaviour and is itself basically deprecated in favour of
+## rlang.
+match_fun_envir <- function(fun, envir = parent.frame(), envir_base = NULL) {
+  if (is_call(fun, quote(quote))) {
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/bulk.R]: `is_call` is just a utility that checks to see if a language object calls a particular function.

double quoted args turned up in user-use (swapping between lapply and lapply_ and not unquoting from memory)

File: R/rrq_controller.R (hunk 1 above)
Comment (richfitz) [reply R/rrq_controller.R]: envir_base is now removed as I can't remember what it was for! Docs have been added

File: R/rrq_controller.R [hunk 4]
```diff
@@ -201,6 +241,99 @@ rrq_controller_ <- R6::R6Class(
       task_submit(self$con, self$keys, dat, key_complete)
     },
 
+    ##' @description Apply a function over a list of of data. This is
```
Comment (r-ash) [inline R/rrq_controller.R:244]:
```suggestion
    ##' @description Apply a function over a list of data. This is
```

**Review by r-ash — APPROVED** [review]
This is v neat, looks good. One tiny typo in suggestions.

**Pull request merged**